flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java

```diff
@@ -0,0 +1,1000 @@
+				FlinkKafkaProducer<byte[], byte[]> producer =
+					initTransactionalProducer(kafkaTransaction.transactionalId, false);
+				producer.resumeTransaction(kafkaTransaction.producerId, kafkaTransaction.epoch);
+				try {
+					producer.commitTransaction();
+					producer.close();
+				}
+				catch (InvalidTxnStateException ex) {
+					// That means we have committed this transaction before.
+					LOG.warn("Encountered error [%s] while recovering transaction [%s]. " +
+						"Presumably this transaction has been already committed before",
+						ex,
+						transaction);
+				}
+				break;
+			case AT_LEAST_ONCE:
+			case NONE:
+				break;
+			default:
+				throw new UnsupportedOperationException("Not implemented semantic");
+		}
+	}
+
+	@Override
+	protected void abort(KafkaTransactionState transaction) {
+		switch (semantic) {
+			case EXACTLY_ONCE:
+				transaction.producer.abortTransaction();
+				producersPool.add(transaction.producer);
+				break;
+			case AT_LEAST_ONCE:
+			case NONE:
+				producersPool.add(transaction.producer);
+				break;
+			default:
+				throw new UnsupportedOperationException("Not implemented semantic");
+		}
+	}
+
+	@Override
+	protected void recoverAndAbort(KafkaTransactionState transaction) {
+		switch (semantic) {
+			case EXACTLY_ONCE:
+				FlinkKafkaProducer<byte[], byte[]> producer =
+					initTransactionalProducer(transaction.transactionalId, false);
+				producer.resumeTransaction(transaction.producerId, transaction.epoch);
+				producer.abortTransaction();
+				producer.close();
+				break;
+			case AT_LEAST_ONCE:
+			case NONE:
+				break;
+			default:
+				throw new UnsupportedOperationException("Not implemented semantic");
+		}
+	}
+
+	private void acknowledgeMessage() {
+		pendingRecords.decrementAndGet();
+	}
+
+	/**
+	 * Flush pending records.
+	 * @param transaction
+	 */
+	private void flush(KafkaTransactionState transaction) throws Exception {
+		if (transaction.producer != null) {
+			transaction.producer.flush();
+		}
+		long pendingRecordsCount = pendingRecords.get();
+		if (pendingRecordsCount != 0) {
+			throw new IllegalStateException("Pending record count must be zero at this point: " + pendingRecordsCount);
+		}
+
+		// if the flushed requests has errors, we should propagate it also and fail the checkpoint
+		checkErroneous();
+	}
+
+	@Override
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
+		super.snapshotState(context);
+
+		transactionalIdsState.clear();
+		for (String transactionalId : availableTransactionalIds) {
+			transactionalIdsState.add(transactionalId);
+		}
+		for (String transactionalId : usedTransactionalIds) {
+			transactionalIdsState.add(transactionalId);
+		}
+	}
+
+	@Override
+	public void initializeState(FunctionInitializationContext context) throws Exception {
+		availableTransactionalIds.clear();
+		for (int i = 0; i < kafkaProducersPoolSize; i++) {
+			availableTransactionalIds.add(UUID.randomUUID().toString());
```

Comment:
That's a valid issue, however on it's one this solution would not be enough. It would not work for a case when we first (1) scale down, then we (2) scale up. On event (2), we would need to create new transactional ids, but we wouldn't know from which id we can start.

However I think we can deduce the starting point for new IDs using `getUnionListState` to track down globally what is the next available transactional id.